### PR TITLE
Add SMI enumeration mode to the smm_ptr module

### PR DESCRIPTION
**Pull request by cpey**
The following pull request extends the `smm_ptr` module by adding an additional
mode called `scan`.

The goal of the `scan` mode is to identify SMIs that trigger time-consuming
code paths by measuring their execution time. This new mode operates similarly
to the fuzz mode and uses a new syscall that returns the SMI execution time.
With the intention to minimize system crashes, its operation bails out from the
currently scanned SMI code when an SMI taking considerably longer is
encountered or memory changes have been detected.

A time-consuming SMI is referred to as 'outlier' in the implementation, and
determined by the configuration option `OUTLIER_THRESHOLD`.

When an 'outlier' is identified, a message similar to the one shown below is
added to the output log.

~~~
[!] SCANNED: SMI# 57 average 785652 checked 1
    Identified outlier: duration 2456476 code 57 data 00 (rax=B284F000 rbx=B284F000 rcx=00, rdx=B284F000 rsi=B284F000 rdi=B284F000)
~~~

The message informs about the average CPU counts for that SMI code, the number
of SMIs executed for that code (scanning for data and rcx values) until the
first outlier is found or a memory change detected (in `checked`), followed by
information about the outlier.

When no outlier is found, the log message does not include the 'outlier'
information.

~~~
[*] SCANNED: SMI# 58 average 759601 checked 4096
~~~

At the end of the report, it lists the number of long-running SMIs found,
followed by the remaining smm_ptr results.

~~~
[+] <<< Done: found 5 long-running SMIs
[-] <<< Done: found 4 potential occurrences of unchecked input pointers

[CHIPSEC] ***************************  SUMMARY  ***************************
[CHIPSEC] Time elapsed            547.306
[CHIPSEC] Modules total           1
[CHIPSEC] Modules failed to run   0:
[CHIPSEC] Modules passed          0:
[CHIPSEC] Modules information     0:
[CHIPSEC] Modules failed          1:
[-] FAILED: chipsec.modules.tools.smm.smm_ptr
[CHIPSEC] Modules with warnings   0:
[CHIPSEC] Modules not applicable  0:
[CHIPSEC] *****************************************************************
~~~

### CODE REVIEW

File: chipsec/modules/tools/smm/smm_ptr.py [hunk 1]
```diff
@@ -338,7 +484,7 @@ def test_config(self, thread_id, _smi_config_fname, _addr, _addr1):
         for line in fcfg:
             if '' == line.strip():
                 self.logger.log(f'\n[*] Testing SMI# 0x{_smi_desc.smi_code:02X} (data: 0x{_smi_desc.smi_data:02X}) {_smi_desc.name} ({_smi_desc.desc})')
-                if self.smi_fuzz_iter(thread_id, _addr, _smi_desc):
+                if selfsmi_fuzz_iter(thread_id, _addr, _smi_desc):
```
Comment (npmitche) [inline chipsec/modules/tools/smm/smm_ptr.py:487]: Looks like you accidentally removed the `.` from this line.

**Review by npmitche — COMMENTED** [review]
At first glance this looks good. We'll need to take some time to test it out. 

Eventually we'll need to handle the case when this is attempted to be run in other environments.

File: chipsec/modules/tools/smm/smm_ptr.py [hunk 2]
```diff
@@ -317,9 +460,12 @@
             if contents_changed:
                 msg = f'DETECTED: SMI# {_smi_desc.smi_code:X} data {_smi_desc.smi_data:X} (rax={_rax:X} rbx={_rbx:X} rcx={_rcx:X} rdx={_rdx:X} rsi={_rsi:X} rdi={_rdi:X})'
                 self.logger.log_important(msg)
-                if FUZZ_BAIL_ON_1ST_DETECT:
+                if FUZZ_BAIL_ON_1ST_DETECT and not scan:
                     raise BadSMIDetected(msg)
 
+        if scan:
+            scan.add(duration, _smi_desc.smi_code, _smi_desc.smi_data, _smi_desc.gprs, contents_changed)
```
[the remark this answers is not in the document]
Comment (cpey) [reply chipsec/modules/tools/smm/smm_ptr.py:467]: This is incorrect. 'duration' is initialized at line 447 when 'scan' is not None.

**Comment by cpey** [comment]
Hi @npmitche, is there any progress on this?

File: chipsec/modules/tools/smm/smm_ptr.py [hunk 3]
```diff
@@ -36,6 +36,7 @@
     * ``config`` = use SMI configuration file <config_file>
     * ``fuzz`` = fuzz all SMI handlers with code in the range <smic_start:smic_end>
     * ``fuzzmore`` = fuzz mode + pass 2nd-order pointers within buffer to SMI handlers
+    * ``scan`` = fuzz mode + time measurement to identify SMIs that trigger long-running code paths
 - ``size``: size of the memory buffer (in Hex)
 - ``address``: physical address of memory buffer to pass in GP regs to SMI handlers (in Hex)
```
Comment (npmitche) [inline chipsec/modules/tools/smm/smm_ptr.py:42]: Will you please add in the "Examples" section (after line 73:
` >>> chipsec_main.py -m tools.smm.smm_ptr -a scan 0x0:0xff`

**Review by npmitche — CHANGES_REQUESTED** [review]

**Review by npmitche — APPROVED** [review]

**Review by Sae86 — APPROVED** [review]

**Pull request merged**